modyn/supervisor/internal/eval/handler.py

```diff
@@ -71,4 +71,55 @@ def get_eval_requests_after_pipeline(self, df_trainings: pd.DataFrame) -> list[E
             set(df_trainings.columns)
         ) == set()
 
-        return []  # followup-PR
+        training_intervals: list[tuple[int, int]] = [
+            (row["first_timestamp"], row["last_timestamp"]) for _, row in df_trainings.iterrows()
+        ]
+        eval_intervals = self.eval_strategy.get_eval_intervals(training_intervals)
+        df_eval_intervals = pd.DataFrame(
+            [
+                (eval_interval.start, eval_interval.end, eval_interval.most_recent_model_interval_end_before)
+                for eval_interval in eval_intervals
+            ],
+            columns=["start", "end", "most_recent_model_interval_end_before"],
+        )
+
+        # now build & potentially reduce the Trainings x EvalIntervals space
+        df_cross = df_eval_intervals.merge(df_trainings, how="cross")
+
+        # to check if a combination is the most recent, we first compute if model was trained before the usage starts
+        # last_timestamp (df_trainings) defines the end of the training data;
+        # most_recent_model_interval_end_before (df_eval_intervals): defines center of an eval intervals.
+        df_cross["chronology_ok"] = df_cross["last_timestamp"] <= df_cross["most_recent_model_interval_end_before"]
+
+        # find the maximum model for every EvalCandidate that doesn't violate that constraint
+        max_model_id = (
```

Comment:
I think, in this instance dataframes are well suited for business logic as their expressiveness is so much higher for these kinds of operations.